File: 2022 November/LivestockLineup.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <string>
#include <vector>

using namespace std;

int cows;
vector<vector<int> > adjList;
map<string, int> alphOrder;
vector<bool> isUsed;
vector<int> cowArrange;

int main(){
    ifstream input("lineup.in");
    ofstream output("lineup.out");
    alphOrder["Beatrice"] = 0;
    alphOrder["Belinda"] = 1;
    alphOrder["Bella"] = 2;
    alphOrder["Bessie"] = 3;
    alphOrder["Betsy"] = 4;
    alphOrder["Blue"] = 5;
    alphOrder["Buttercup"] = 6;
    alphOrder["Sue"] = 7;

    input >> cows;
    isUsed.resize(8, false);
    adjList.resize(8);

    for(int i = 0; i < cows; i++){
        string a, trash, f;
        input >> a >> trash >> trash >> trash >> trash >> f;
        // cout << a << " " << f << endl;
        adjList[alphOrder[a]].push_back(alphOrder[f]);
        adjList[alphOrder[f]].push_back(alphOrder[a]);
        // cout << adjList[alphOrder[a]][0] << " " << adjList[alphOrder[f]][0] << endl;
    }
    
    // cout << "done";

    for(int i = 0; i < 8; i++){
        // cout << i;
        if(!isUsed[i] && adjList[i].size() == 0){
            cowArrange.push_back(i);
            isUsed[i] = true;
        }

        if(adjList[i].size() == 1 && !isUsed[i]){
            cowArrange.push_back(i);
            isUsed[i] = true;
            int x = i;
            while(adjList[x].size() > 0){
            
                if(adjList[x].size() == 1){
                    if(!isUsed[adjList[i][0]]){
                        cowArrange.push_back(adjList[x][0]);
                        isUsed[adjList[x][0]] = true;
                        x = adjList[x][0];
                        continue;
                    } else {
                        break;
                    }
                }

                if(adjList[x].size() == 2){
                    if(!isUsed[adjList[x][0]]){
                        cowArrange.push_back(adjList[x][0]);
                        isUsed[adjList[x][0]] = true;
                        x = adjList[x][0];
                    } else if (!isUsed[adjList[x][1]]){
                        cowArrange.push_back(adjList[x][1]);
                        isUsed[adjList[x][1]] = true;
                        x = adjList[x][1];
                    } else {
                        break;
                    }
                }
            }
        }
    }
    for(int i = 0; i < 8; i++){
        // cout << cowArrange[i] << " ";
        for(auto const& [key, val] : alphOrder){
            if(val == cowArrange[i]){
                output << key << endl;
                break;
            }
        }
    }
}

File: 2023 March/livestockLineup.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;



public class livestockLineup {

    private static ArrayList<Integer> inComp;
    private static boolean[] visited = new boolean[8];
    private static ArrayList<Integer>[] adjMatrix = new ArrayList[8];


    private static void DFS(int a){
        if(visited[a]){
            return;
        }
        System.out.println(a);
        visited[a] = true;
        inComp.add(a);
        for(int i = 0; i < adjMatrix[a].size(); i++){
            DFS(adjMatrix[a].get(i));
        }
        
    }

    public static void main(String[] args){
        Scanner input = new Scanner(System.in);
        int constraints = Integer.parseInt(input.next());
        String[] cowNames = {"Beatrice", "Sue", "Belinda", "Bessie", "Betsy", "Blue", "Bella", "Buttercup"};
        Arrays.sort(cowNames);
        Map<String, Integer> alphOrder = new HashMap<String, Integer>();
        for(int i = 0; i < cowNames.length; i++){
            alphOrder.put(cowNames[i], i);
        }

        for(int i = 0; i < adjMatrix.length; i++){
            adjMatrix[i] = new ArrayList<Integer>();
        }

        while(input.hasNextLine()){
            int cow1, cow2;
            cow1 = alphOrder.get(input.next()); input.next(); input.next(); input.next(); input.next(); cow2 = alphOrder.get(input.next());
            System.out.println(cow1 + " " + cow2);
            adjMatrix[cow1].add(cow2);
            adjMatrix[cow2].add(cow1);
        }

        List<ArrayList<Integer> > components = new ArrayList<ArrayList<Integer> >();
        for(int i = 0; i < visited.length; i++){
            if(!visited[i]){
                inComp = new ArrayList<Integer>();
                DFS(i);
                for(int j = 0; j < inComp.size(); i++){
                    System.out.print(inComp.get(j) + " ");
                }
                System.out.println();
                components.add(inComp);
            }
        }
        components.sort(new Comparator<ArrayList<Integer>>() {    
            @Override
            public int compare(ArrayList<Integer> o1, ArrayList<Integer> o2) {
                return o1.get(0).compareTo(o2.get(0));
            }               
        });
        for(int i = 0; i < components.size(); i++){
            for(int j = 0; j < components.get(i).size(); i++){
                System.out.println(components.get(i).get(j));
            }
        }
        
    }
}